ad7689, ad7682, ad7949, ad7699 driver

buha opened the pull request:
Initial version of the driver for these four PulSAR ADC's.
The driver uses the Read After Conversion (RAC) mode for
communication with the device.
This could be improved by using the Read Spanning Conversion
(RSC) mode to allow even slower hosts to be able to retrieve
data at maximum sample rate.
However, RAC was chosen due to it being standard and easy to
implement on any platform (bare-metal microcontrollers, SoC's
running linux etc.) that no-OS can run on.

Signed-off-by: Darius Berghe <darius.berghe@analog.com>

amiclaus commented on drivers/adc/ad7689/ad7689.c:
```diff
@@ -0,0 +1,194 @@
+#include <stdlib.h>
```
file header comment missing.

amiclaus replied:
Also header comments for each function should be added.

buha replied:
you mean doxygen documentation ?

amiclaus replied:
yes

buha replied:
just added doxygen

amiclaus commented on drivers/adc/ad7689/ad7689.c:
```diff
@@ -0,0 +1,194 @@
+	free(dev);
+
+	return SUCCESS;
+}
```
nitpick: new line at end of file

buha replied:
done

amiclaus commented on drivers/adc/ad7689/ad7689.c:
```diff
@@ -0,0 +1,194 @@
+#include "print_log.h"
+#include "delay.h"
+
+const char *ad7689_device_name[] = {
```
maybe add an attribute `name` to the device descriptor? (just a suggestion)
Some examples: 
https://github.com/analogdevicesinc/no-OS/blob/7c63e0dd4d5342106e0c7d33205ebdb5db5f5c2b/drivers/axi_core/jesd204/axi_jesd204_rx.h#L78-L79

https://github.com/analogdevicesinc/no-OS/blob/7c63e0dd4d5342106e0c7d33205ebdb5db5f5c2b/drivers/axi_core/jesd204/axi_adxcvr.h#L53

https://github.com/analogdevicesinc/no-OS/blob/7c63e0dd4d5342106e0c7d33205ebdb5db5f5c2b/drivers/axi_core/axi_dac_core/axi_dac_core.h#L58-L59

buha replied:
yeah, this would be better

amiclaus commented on drivers/adc/ad7689/ad7689.h:
```diff
@@ -0,0 +1,136 @@
+		    uint32_t nb_samples);
+int32_t ad7689_remove(struct ad7689_dev *dev);
+
+#endif
```
nitpick: new line at the end

buha commented:
V2:
- added doxygen
- inverted c->rb logic as it was counterintuitive (c->rb = false previously meant readback is active)
- simplified 16->14 bits scaling with just one common signed division by four instead of checking for bipolar/unipolar setting
- addressed the comments above

buha commented:
V3:
- fixed cppcheck

cristiantpop commented on drivers/adc/ad7689/ad7689.c:
```diff
@@ -0,0 +1,291 @@
+	// write config twice to also perform two dummy conversions
+	ret = ad7689_write_config(d, &init_param->config);
+	if (ret < 0)
+		goto error;
```
Call `spi_remove` in this case

buha replied:
code goes to error: label where spi_remove is called in the _remove function

buha replied:
I added a patch to address this

amiclaus commented on drivers/adc/ad7689/ad7689.c:
```diff
@@ -0,0 +1,291 @@
+
+	ret = spi_init(&d->spi_desc, &init_param->spi_init);
+	if (ret < 0)
+		goto error;
```
is it required to call `spi_remove` in this case? or just `free(d)`?

buha replied:
I added a patch to address this

buha commented:
V4: 
- deinit only the necessary stuff on _init failures

amiclaus reviewed (approved):
Perform a rebase with the latest master branch version. Should fix the remaining build issues.
I saw that some of static functions in the no-OS repository are also provided with header comments (so I guess this is up to you if you want to add them here also, or not).
 The rest LGTM.

buha commented:
V5:
- rebased on master / retrigger checks
